### ISSUE TITLE
Kucoin: watchMytrades doesn't work with method "/spot/tradeFills"

### ISSUE THREAD

**Issue by mjwvb** [ccxt/ccxt#23302]
### Operating System

MacOS

### Programming Languages

JavaScript

### CCXT Version

4.3.70

### Description

Related to #23057 when you've build in an extra method param to watchMyTrades() to support the undocumented kucoin "/spot/tradeFills" websocket endpoint. 
However when I use that method I simply don't receive any incoming trades, see code below.

Also, when that method does work, I suspect it won't work correctly anyway. The parseWsTrade() function is used to parse trade info for both websocket endpoints, but it always uses matchPrice/matchSize. matchPrice/matchSize are not in the /spot/tradeFills trade response and should be price/size in that case. Maybe you can look at that as well.

For now I'm still rolling back to 4.3.53 as I need the fee information of kucoin trades using the "/spot/tradeFills" websocket endpoint.

### Code

```typescript
import * as ccxt from "ccxt";

const test = async () => {
  const exchange = new ccxt.pro.kucoin({
    apiKey: "...",
    secret: "...",
    password: "...",
  });

  const watchMyTrades = async () => {
    while (true) {
      try {
        console.log("Watching my trades");
        const trade = await exchange.watchMyTrades(undefined, undefined, undefined, {
          method: "/spot/tradeFills",
        });

        console.log("Received trade:", trade);
      } catch (e) {
        console.log(e);
        if (e instanceof ccxt.BadRequest) throw e;
      }
    }
  };

  watchMyTrades();

  // Wait 5 seconds for the websocket to initialize
  setTimeout(async () => {
    const order = await exchange.createOrder("USDT/USDC", "market", "sell", 1);
    console.log("Placed order:", order);
  }, 5000);
};

test();
```

**Result:** 
No received trade when using the method "/spot/tradeFills". 
When uncommenting the method param I do receive the trade as expected.

**Comment by sc0Vu** [comment]
Hi @mjwvb,

I can receive trade data from both topics, was your order traded? Also you're right about the price and amount when listen to `/spot/tradeFills`, we'll fix this asap.

**Comment by mjwvb** [comment]
@sc0Vu It's really strange. Whatever I do, I don't receive messages from /spot/tradeFills, but I do receive the trade message from the new method... 

When I rollback to 4.3.53 which is based on /spot/tradeFills I also don't receive the trade message. So maybe I'm doing something in the test environment/script as 4.3.53 actually works for my live trading bot...

Thanks for patching parseWsTrade though!

**Comment by mjwvb** [comment]
Executed the following again with 4.3.53 and placed an order on BTC/USDC using the Kucoin UI. As you can see in the verbose log only the order info is coming in and not the trade fills. Is there something wrong with my watchMyTrades() test implementation?

### Code
```typescript
import * as ccxt from "ccxt";

const test = async () => {
  const exchange = new ccxt.pro.kucoin({
    apiKey: "...",
    secret: "...",
    password: "...",
    verbose: true,
  });

  const watchOrders = async () => {
    console.log("Watching orders");
    while (true) {
      try {
        const order = await exchange.watchOrders();
        console.log("Received order:", order);
      } catch (e) {
        console.log(e);
        if (e instanceof ccxt.BadRequest) throw e;
      }
    }
  };

  const watchMyTrades = async () => {
    console.log("Watching my trades");
    while (true) {
      try {
        const trade = await exchange.watchMyTrades();
        console.log("Received trade:", trade);
      } catch (e) {
        console.log(e);
        if (e instanceof ccxt.BadRequest) throw e;
      }
    }
  };

  watchOrders();
  watchMyTrades();
};

test();

```
### Log result
```
2024-08-02T13:07:12.439Z connecting to wss://ws-api-spot.kucoin.com/?token=...&privateChannel=true&connectId=private
2024-08-02T13:07:13.135Z onUpgrade
2024-08-02T13:07:13.137Z onOpen
2024-08-02T13:07:13.143Z onMessage { id: 'private', type: 'welcome' }
2024-08-02T13:07:13.144Z sending {
  id: '1',
  type: 'subscribe',
  topic: '/spotMarket/tradeOrders',
  response: true,
  privateChannel: true
}
2024-08-02T13:07:13.145Z sending {
  id: '2',
  type: 'subscribe',
  topic: '/spot/tradeFills',
  response: true,
  privateChannel: true
}
2024-08-02T13:07:13.373Z onMessage { id: '1', type: 'ack' }
2024-08-02T13:07:13.377Z onMessage { id: '2', type: 'ack' }
2024-08-02T13:07:29.853Z onMessage {
  topic: '/spotMarket/tradeOrders',
  type: 'message',
  data: {
    feeType: 'takerFee',
    filledSize: '0.00012474',
    funds: '8.0496',
    liquidity: 'taker',
    matchPrice: '64526.5',
    matchSize: '0.00012474',
    orderId: '66acda1130f3dd0007025f54',
    orderTime: 1722604049761,
    orderType: 'market',
    remainFunds: '0.00056439',
    side: 'buy',
    status: 'match',
    symbol: 'BTC-USDC',
    tradeId: '7490726501961729',
    ts: '1722604049763000000',
    type: 'match'
  },
  subject: 'orderChange',
  userId: '...',
  channelType: 'private'
}
2024-08-02T13:07:29.868Z onMessage {
  topic: '/spotMarket/tradeOrders',
  type: 'message',
  data: {
    filledSize: '0.00012474',
    funds: '8.0496',
    orderId: '66acda1130f3dd0007025f54',
    orderTime: 1722604049761,
    orderType: 'market',
    remainFunds: '0',
    remainSize: '0',
    side: 'buy',
    status: 'done',
    symbol: 'BTC-USDC',
    ts: '1722604049763000000',
    type: 'canceled'
  },
  subject: 'orderChange',
  userId: '...',
  channelType: 'private'
}
Received order: [
  {
    info: {
      filledSize: '0.00012474',
      funds: '8.0496',
      orderId: '66acda1130f3dd0007025f54',
      orderTime: 1722604049761,
      orderType: 'market',
      remainFunds: '0',
      remainSize: '0',
      side: 'buy',
      status: 'done',
      symbol: 'BTC-USDC',
      ts: '1722604049763000000',
      type: 'canceled'
    },
    symbol: 'BTC/USDC',
    id: '66acda1130f3dd0007025f54',
    clientOrderId: undefined,
    timestamp: 1722604049761,
    datetime: '2024-08-02T13:07:29.761Z',
    lastTradeTimestamp: undefined,
    type: 'market',
    timeInForce: 'IOC',
    postOnly: undefined,
    side: 'buy',
    price: undefined,
    stopPrice: undefined,
    triggerPrice: undefined,
    amount: undefined,
    cost: undefined,
    average: undefined,
    filled: 0.00012474,
    remaining: undefined,
    status: 'canceled',
    fee: undefined,
    trades: [],
    fees: [],
    lastUpdateTimestamp: undefined,
    reduceOnly: undefined,
    takeProfitPrice: undefined,
    stopLossPrice: undefined
  }
]
2024-08-02T13:07:31.140Z sending { id: '4', type: 'ping' }
2024-08-02T13:07:31.364Z onMessage { id: '4', type: 'pong' }
2024-08-02T13:07:49.142Z sending { id: '5', type: 'ping' }
2024-08-02T13:07:49.397Z onMessage { id: '5', type: 'pong' }
```

**Comment by carlosmiei** [comment]
@mjwvb Running your script I still get trades through watchMyTrades with the `spot/TradeFills` channel, log

```

2024-08-02T14:21:34.291Z connecting to wss://ws-api-spot.kucoin.com/?token=2neAiuYvAU737TOajb2U3uT8AEZqSWYe0fBD4LoHuXJDSC7gIzJiH4kNTWhCPISWo6nDpAe7aUaXDxagG5mJ12KN1Vw2-CvAqM3YlnG2CCneQJPw53OOlnwZIngtMdMrjqPnP-biofFuXcsXSiYV2L7A_0jPjNB1JBvJHl5Vs9Y%3D.ESpO5UtqC0tLHgBqWFeLXA%3D%3D&privateChannel=true&connectId=private
2024-08-02T14:21:35.111Z onUpgrade
2024-08-02T14:21:35.114Z onOpen
2024-08-02T14:21:35.115Z onMessage { id: 'private', type: 'welcome' }
2024-08-02T14:21:35.116Z sending {
  id: '1',
  type: 'subscribe',
  topic: '/spotMarket/tradeOrders',
  response: true,
  privateChannel: true
}
2024-08-02T14:21:35.117Z sending {
  id: '2',
  type: 'subscribe',
  topic: '/spot/tradeFills',
  response: true,
  privateChannel: true
}
2024-08-02T14:21:35.367Z onMessage { id: '1', type: 'ack' }
2024-08-02T14:21:35.373Z onMessage { id: '2', type: 'ack' }
2024-08-02T14:21:42.328Z onMessage {
  topic: '/spotMarket/tradeOrders',
  type: 'message',
  data: {
    clientOid: 'fa64811f-8354-43e3-b3af-1ef9d120d872',
    feeType: 'takerFee',
    filledSize: '0.1',
    funds: '7.160701',
    liquidity: 'taker',
    matchPrice: '67.881',
    matchSize: '0.1',
    orderId: '66aceb762555b80007716ff2',
    orderTime: 1722608502269,
    orderType: 'market',
    remainFunds: '0.372601',
    remainSize: '0',
    side: 'buy',
    size: '0.1',
    status: 'match',
    symbol: 'LTC-USDT',
    tradeId: '7491981477826561',
    ts: '1722608502289000000',
    type: 'match'
  },
  subject: 'orderChange',
  userId: '64f99aced178640001306e6e',
  channelType: 'private'
}
2024-08-02T14:21:42.335Z onMessage {
  topic: '/spotMarket/tradeOrders',
  type: 'message',
  data: {
    clientOid: 'fa64811f-8354-43e3-b3af-1ef9d120d872',
    filledSize: '0.1',
    funds: '7.160701',
    orderId: '66aceb762555b80007716ff2',
    orderTime: 1722608502269,
    orderType: 'market',
    remainFunds: '0',
    remainSize: '0',
    side: 'buy',
    size: '0.1',
    status: 'done',
    symbol: 'LTC-USDT',
    ts: '1722608502289000000',
    type: 'filled'
  },
  subject: 'orderChange',
  userId: '64f99aced178640001306e6e',
  channelType: 'private'
}
Received trade: [
  {
    info: {
      clientOid: 'fa64811f-8354-43e3-b3af-1ef9d120d872',
      feeType: 'takerFee',
      filledSize: '0.1',
      funds: '7.160701',
      liquidity: 'taker',
      matchPrice: '67.881',
      matchSize: '0.1',
      orderId: '66aceb762555b80007716ff2',
      orderTime: 1722608502269,
      orderType: 'market',
      remainFunds: '0.372601',
      remainSize: '0',
      side: 'buy',
      size: '0.1',
      status: 'match',
      symbol: 'LTC-USDT',
      tradeId: '7491981477826561',
      ts: '1722608502289000000',
      type: 'match'
    },
    timestamp: 1722608502288,
    datetime: '2024-08-02T14:21:42.288Z',
    symbol: 'LTC/USDT',
    id: '7491981477826561',
    order: '66aceb762555b80007716ff2',
    type: 'market',
    takerOrMaker: 'taker',
    side: 'buy',
    price: 67.881,
    amount: 0.1,
    cost: 6.7881,
    fee: { cost: undefined, rate: undefined, currency: 'USDT' },
    fees: [ [Object] ]
  }
]
Watching my trades
Received order: [
  {
    info: {
      clientOid: 'fa64811f-8354-43e3-b3af-1ef9d120d872',
      filledSize: '0.1',
      funds: '7.160701',
      orderId: '66aceb762555b80007716ff2',
      orderTime: 1722608502269,
      orderType: 'market',
      remainFunds: '0',
      remainSize: '0',
      side: 'buy',
      size: '0.1',
      status: 'done',
      symbol: 'LTC-USDT',
      ts: '1722608502289000000',
      type: 'filled'
    },
    symbol: 'LTC/USDT',
    id: '66aceb762555b80007716ff2',
    clientOrderId: 'fa64811f-8354-43e3-b3af-1ef9d120d872',
    timestamp: 1722608502269,
    datetime: '2024-08-02T14:21:42.269Z',
    lastTradeTimestamp: undefined,
    type: 'market',
    timeInForce: 'IOC',
    postOnly: undefined,
    side: 'buy',
    price: undefined,
    stopPrice: undefined,
    triggerPrice: undefined,
    amount: 0.1,
    cost: undefined,
    average: undefined,
    filled: 0.1,
    remaining: 0,
    status: 'closed',
    fee: undefined,
    trades: [],
    fees: [],
    lastUpdateTimestamp: undefined,
    reduceOnly: undefined,
    takeProfitPrice: undefined,
    stopLossPrice: undefined
  }
]
2024-08-02T14:21:42.392Z onMessage {
  type: 'message',
  topic: '/spot/tradeFills',
  userId: '64f99aced178640001306e6e',
  channelType: 'private',
  subject: '/spot/tradeFills',
  data: {
    fee: 0.0067881,
    feeCurrency: 'USDT',
    feeRate: 0.001,
    orderId: '66aceb762555b80007716ff2',
    orderType: 'market',
    price: 67.881,
    side: 'buy',
    size: 0.1,
    symbol: 'LTC-USDT',
    time: '1722608502289000000',
    tradeId: '7491981477826561'
  }
}
Received trade: [
  {
    info: {
      fee: 0.0067881,
      feeCurrency: 'USDT',
      feeRate: 0.001,
      orderId: '66aceb762555b80007716ff2',
      orderType: 'market',
      price: 67.881,
      side: 'buy',
      size: 0.1,
      symbol: 'LTC-USDT',
      time: '1722608502289000000',
      tradeId: '7491981477826561'
    },
    timestamp: 1722608502288,
    datetime: '2024-08-02T14:21:42.288Z',
    symbol: 'LTC/USDT',
    id: '7491981477826561',
    order: '66aceb762555b80007716ff2',
    type: 'market',
    takerOrMaker: undefined,
    side: 'buy',
    price: 67.881,
    amount: 0.1,
    cost: 6.7881,
    fee: { cost: 0.0067881, rate: 0.001, currency: 'USDT' },
    fees: [ [Object] ]
  }
]
```

**Comment by mjwvb** [comment]
@carlosmiei thanks for testing!

I really don't get it. Tested it on node v20 with ccxt v4.3.53 and v4.3.73 with the method param. Whatever I do I only receive the order events and never the events from /spot/tradeFills. 

What boggles me the most is that my live trading bot seems to be receiving the /spot/tradeFills events just fine, or at least according to my trading logs that is... So I tested this script above with that same api key on that same live environment: again no event messages. So now I'm not sure anymore.

What might be the cause on my side? I am using the API V3, don't know if that's a thing? Seems to be the default for a long enough time though.

**Comment by mjwvb** [comment]
Had some extra logging turned on last night. My live trading bot really seems to be receiving all trades correctly using 4.3.53's /spot/tradeFills. I'm gonna assume it works correctly then, and just accept there's mystery in life.

0|npm run  | receiveTrade {
0|npm run  |   trade: {
0|npm run  |     info: {
0|npm run  |       fee: 1.542260588448,
0|npm run  |       feeCurrency: 'USDT',
0|npm run  |       feeRate: 0.001,
0|npm run  |       orderId: '66aff28df9c24700079bdf46',
0|npm run  |       orderType: 'market',
0|npm run  |       price: 58516.8,
0|npm run  |       side: 'buy',
0|npm run  |       size: 0.02635586,
0|npm run  |       symbol: 'BTC-USDT',
0|npm run  |       time: '1722806925077000000',
0|npm run  |       tradeId: '9931725505314817',
0|npm run  |       parseTimestamp: 1722806925120
0|npm run  |     },
0|npm run  |     timestamp: 1722806925076,
0|npm run  |     datetime: '2024-08-04T21:28:45.076Z',
0|npm run  |     symbol: 'BTC/USDT',
0|npm run  |     id: '9931725505314817',
0|npm run  |     order: '66aff28df9c24700079bdf46',
0|npm run  |     type: 'market',
0|npm run  |     takerOrMaker: undefined,
0|npm run  |     side: 'buy',
0|npm run  |     price: 58516.8,
0|npm run  |     amount: 0.02635586,
0|npm run  |     cost: 1542.260588448,
0|npm run  |     fee: { cost: 1.542260588448, rate: 0.001, currency: 'USDT' },
0|npm run  |     fees: [ [Object] ]
0|npm run  |   }
0|npm run  | }

**Comment by WillPCowan** [comment]
@mjwvb did you get any further with this?
@carlosmiei I am also experiencing this issue. I can confirm that ccxt.pro.kucoin class is receiving trades and processing them in `handle_trade`, but they are never getting passed back from `watch_my_trades`... This used to work but recently hasn't, wondering if it is my application code but struggling to isolate what may be the cause.

I will do a bit more debugging and see if I can find out why.

FYI I stream both watch_orders and watch_my_trades simultaneously, if relevant.

**Comment by WillPCowan** [comment]
Do we have to paramaterize ccxt.pro.kucoin to use '/spot/tradeFills' instead of '/spotMarket/tradeOrders'? Like `exchange['options']['watchMyTrades']['/spot/tradeFills']`?

**Comment by WillPCowan** [comment]
Hey @carlosmiei =) @mjwvb is right. No trade messages are received from the '/spot/tradeFills' channel -- in the new version even updating exchange client options does not result in sending subscription request such as `{'id': '3', 'type': 'subscribe', 'topic': '/spot/tradeFills', 'response': True, 'privateChannel': True}`.

**Comment by mjwvb** [comment]
I went through my trading bot logs and I found out that /spot/tradeFills only works for (cross) margin trades, and not for normal spot trades. I can see it worked perfectly fine for normal spot pairs until somewhere between the 3rd and 5th of June. I didn't do any updates of my trading bot code between those dates, so I think Kucoin must have disabled this websocket endpoint, but only for non-margin spot...

I can now easily reproduce it using the test script above that I receive fill info on BTC/USDT on cross margin, but not on non-margin. Can someone else confirm this, just to be sure? @WillPCowan @sc0Vu @carlosmiei

For now I think I'm going to calculate the fee information manually on the new websocket endpoint using the maker/taker info and the fee rate info of that trading paire. There's no other way to make it reliable besides doing an extra request to get the recent trade fills etc., which I'm not a fan of due to extra latency...
